### DISCUSSION TITLE
Does the `Browse or upload datasets` button need fixes?

### DISCUSSION THREAD

**Discussion by bernt-matthias** [galaxyproject/galaxy#19275] [General Discussion]
The `data dialog` has been [added](https://github.com/galaxyproject/galaxy/pull/7553) to  select datasets from a collections and data libraries with a few clicks less. Unfortunately this bypasses the filters of the parameter and users can select datasets of the wrong format or invalid datasets. In such cases the invalid choice is flagged with the prefix `(Unavailable)` - where I never knew what it means. 

This leads now to the unfortunate situation that tool developers start develop add validators as well as filters to their tools in order to ensure that users really can't feed invalid input to the tools: e.g. [here](https://github.com/qiime2/q2galaxy/issues/60#issuecomment-2520950661). I'm really wondering now what the suggestions to tool developers are: in which cases to add validators / filters / both? The biggest difference between filters and validators seem to be when they are called and performance implications, since filters might need to run over all elements of a history and validators only on the selected item (I'm still not entirely sure when exactly they run: at selection, at tool submission?).

Also filters and validators are IMO actually quite similar [anyway](https://github.com/galaxyproject/galaxy/pull/18407) and I remember that I had a hard time to distinguish them when I started to develop tools. 

Suggestions: 

- Change the prefix `(Unavailable)` to something like `(Invalid choice)`
- Document to tool developers the advantages / disadvantages of validators and filters
- Add filters to the dialog (would probably be to imperformant)?
- Remove the data dialog? Maybe we have nowadays other/better possibilities: easier extraction of data sets from collection and even drag and drop (is this validated)?

**Comment by mvdbeek** [comment]
> Unfortunately this bypasses the filters of the parameter and users can select datasets of the wrong format or invalid datasets.

When you say format and invalid, do you mean wrong datatype or datasets in terminal non-OK states ? That would be bugs.

I assume you mean we don't run filters or validators. The data dialog doesn't work with datasets but files and folders, so I don't think there's much to be done ?

**Reply by bernt-matthias** [reply]
> When you say format and invalid, do you mean wrong datatype or datasets in terminal non-OK states ? That would be bugs.

Yes. In my test I could select a `txt` dataset for a `qza` data parameter and qza that does not pass the filters. But due to the additional validators I could not run the tool (the tool implements filters and validators that do actually the same).
I did not test terminal non-OK states. 

> The data dialog doesn't work with datasets but files and folders, so I don't think there's much to be done ?

I mean this thing with the three dots: 

![Screenshot from 2024-12-06 09-31-51](https://github.com/user-attachments/assets/7e597c79-8d1b-4fa4-8566-7d3a7b77d7cf) 

It allows me to select datasets from my history ...?
